### PR TITLE
UI, controller and suite.xml code for case search

### PR DESCRIPTION
**Pull request by kaapstorm**
This PR replaces the original review PR #11050 , now based off `master`. 

Don't merge yet.

This PR has grown to be ridiculously big. I'll break it into bite-size chunks when it's ready for merging. I still need to add some tests, and pull out commits that are kind of stand-alone, like ba219763

@proteusvacuum cc @millerdev

### CODE REVIEW

File: corehq/apps/ota/tests/test_search_claim_endpoints.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from django.core.urlresolvers import reverse
+from django.test import TestCase, Client
+
+from casexml.apps.case.mock import CaseBlock
+from casexml.apps.case.util import post_case_blocks
+from corehq.apps.case_search.models import CLAIM_CASE_TYPE
+from corehq.apps.domain.shortcuts import create_domain
+from corehq.apps.users.models import CommCareUser
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.form_processor.tests import run_with_all_backends
```
Comment (dimagimon) [inline corehq/apps/ota/tests/test_search_claim_endpoints.py]: F401 'run_with_all_backends' imported but unused

File: corehq/apps/ota/views.py [hunk 2]
```diff
@@ -1,18 +1,30 @@
...
+from corehq.apps.es.case_search import CaseSearchES, flatten_result
 from corehq.apps.ota.forms import PrimeRestoreCacheForm, AdvancedPrimeRestoreCacheForm
 from corehq.apps.ota.tasks import prime_restore
 from corehq.apps.style.views import BaseB3SectionPageView
 from corehq.apps.users.models import CouchUser, CommCareUser
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.form_processor.serializers import CommCareCaseSQLSerializer, get_instance_from_data
+from corehq.pillows.const import MAX_SEARCH_RESULTS
 from corehq.tabs.tabclasses import ProjectSettingsTab
+from corehq.form_processor.models import CommCareCaseSQL
```
Comment (dimagimon) [inline corehq/apps/ota/views.py]: F401 'CommCareCaseSQL' imported but unused

File: corehq/apps/domain/views.py [hunk 3]
```diff
@@ -2071,6 +2080,59 @@ def post(self, request, *args, **kwargs):
...
+            for key in query_dict:
+                match = pattern.match(key)
+                if match:
+                    i = int(match.group('index'))
+                    attr = match.group('attr')
+                    is_list = key.endswith('[]')  # i.e. "...[properties][]"
+                    fuzzy_dict[i][attr] = query_dict.getlist(key) if is_list else query_dict[key]
+            if not fuzzy_dict:
+                return []
+            return [fuzzy_dict[i] for i in range(max(fuzzy_dict.keys()) + 1) if fuzzy_dict[i]]
```
Comment (dimagimon) [inline corehq/apps/domain/views.py]: F812 list comprehension redefines 'i' from line 2107

**Comment by kaapstorm** [comment]
@proteusvacuum not sure what I've added since you last glanced. Maybe :tropical_fish: from "No point in returning more than 10 results" (f17471b). Recent ones are just bug fixes.

File: corehq/apps/app_manager/tests/test_suite_sync_request.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+from django.test import SimpleTestCase
+from mock import patch
+
+from corehq.apps.app_manager.const import APP_V2
+from corehq.apps.app_manager.models import Application, Module, CaseSearch, CaseSearchProperty, UpdateCaseAction
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_suite_sync_request.py]: F401 'UpdateCaseAction' imported but unused

File: corehq/apps/ota/views.py [hunk 5]
```diff
@@ -81,28 +77,8 @@ def claim(request, domain):
...
-                'host': IndexAttrs(
-                    case_type=host_type,
-                    case_id=host_id,
-                    relationship='extension',
-                )
-            }
-        ).as_xml()
-        post_case_blocks([claim_case_block], {'domain': domain})
+        claim_case(domain, couch_user.user_id, request.POST['case_id'],
+                   host_type=request.POST.get('case_type'), host_name=request.POST.get('case_name'))
```
Comment (proteusvacuum) [inline corehq/apps/ota/views.py]: :+1:

File: corehq/apps/domain/views.py [hunk 6]
```diff
@@ -2097,23 +2097,25 @@ def unpack_fuzzies(query_dict):
...
-                if match:  # non-fuzzy-properties won't match, i.e. "^enable$"
-                    fuzzy_dict[int(match.group('index'))][match.group('attr')] = value
+                if match:
+                    i = int(match.group('index'))
+                    attr = match.group('attr')
+                    is_list = key.endswith('[]')  # i.e. "...[properties][]"
+                    fuzzy_dict[i][attr] = query_dict.getlist(key) if is_list else query_dict[key]
             if not fuzzy_dict:
                 return []
             return [fuzzy_dict[i] for i in range(max(fuzzy_dict.keys()) + 1) if fuzzy_dict[i]]
```
Comment (proteusvacuum) [inline corehq/apps/domain/views.py]: I think you could achieve the same thing here more pythonically with `[item for key, item in fuzzy_dict.iter_items() if item]`, unless I've misunderstood what this is doing

**Comment by proteusvacuum** [comment]
Looks good Norman. Build fail, and conflicts, but looking forward to getting this in!
```
======================================================================

ERROR: test_case_list_form_media (corehq.apps.app_manager.tests.test_media_suite.LocalizedMediaSuiteTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/mnt/corehq/apps/app_manager/tests/test_media_suite.py", line 228, in test_case_list_form_media

    self._assert_app_strings_available(app, 'en')

  File "/mnt/corehq/apps/app_manager/tests/test_media_suite.py", line 266, in _assert_app_strings_available

    app_strings = commcare_translations.loads(app.create_app_strings(lang))

  File "/mnt/corehq/apps/app_manager/models.py", line 4744, in create_app_strings

    return gen.create_app_strings(self, lang)

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 161, in create_app_strings

    for part in self.app_strings_parts(app, lang, for_default=for_default):

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 250, in app_strings_parts

    for_default=for_default)

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 154, in create_custom_app_strings

    custom = dict(_create_custom_app_strings(app, lang, for_default=for_default))

  File "/mnt/corehq/apps/app_manager/app_strings.py", line 115, in _create_custom_app_strings

    yield id_strings.search_property_locale(module, prop.name), trans(prop.label)

AttributeError: 'str' object has no attribute 'name'
```
(Same error in other build too)

File: corehq/apps/domain/views.py (hunk 6 above)
Comment (kaapstorm) [inline corehq/apps/domain/views.py]: Hmmm. Not sure why I did it that way -- maybe a little too focused on that index. Could even just `[item for item in fuzzy_dict.iter_values() if item]` or `filter(lambda x: x, fuzzy_dict.values())`. Thank you.

**Comment by proteusvacuum** [comment]
:boot: :construction_worker:

**Comment by kaapstorm** [comment]
:white_check_mark: 

Three things to note about this branch that I would like to add in the very near future, and all pretty small:

1. Save button on page to enable search and configure fuzzy properties, under Project Settings, isn't notified of changes. It is always active. Would be nice to enable after change, and disable after save.
2. When claiming a case, session doesn't store whether the user has already just claimed the case, as per spec. This means that multiple claims can be created for the same case by the same user. I should prevent this, and add a test for it.
3. Module case list page should allow user to change wording of the "Search" command. suite.xml allows for it, but UI doesn't offer it.

Dropping "Hold off on merge" but noting that these three things would be very nice, and (2) is required.

**Comment by proteusvacuum** [comment]
:tada:

**Pull request merged**